[BUGFIX release] `willRender` is called for `attributeBindings` changes.

rwjblue opened the pull request:
Depending on which node-manager was used, `scope.view` or `scope.component` might be used (but not both). For `attributeBindings` case `scope.component` was not set, therefore the various hooks were not being called properly.

rwjblue committed:
[BUGFIX release] `willRender` is called for `attributeBindings` changes.

Depending on which node-manager was used, `scope.view` or
`scope.component` might be used (but not both). For
`attributeBindings` case `scope.component` was not set, therefore the
various hooks were not being called properly.

tomdale commented on packages/ember-htmlbars/tests/integration/component_lifecycle_test.js:
```diff
@@ -344,6 +302,51 @@ QUnit.test('changing a component\'s displayed properties inside didInsertElement
+    template: compile('{{the-top}}'),
+    container: container
+  }).create();
+
```
This test is testing components, so why is `scope.component` not set? Surely it goes through the `ComponentNodeManager`? Am I just misunderstanding the test?

rwjblue replied:
I agree that this seems somewhat fishy.  It is possible that this fix is still correct, but I'd like to figure out why `scope.component` is not set before we decide to pull the trigger on it.

jmurphyau commented:
Does this fix #11533? I know its already fixed with another PR but this I think could have also fixed it

rwjblue commented:
@jmurphyau - Yes, that is exactly what made me start thinking about this.

rwjblue commented:
@tomdale - I tracked down why `scope.component` is not set:

* `scope.component` should be set in `bind-shadow-scope`, but is only set when `options.attrs` and `options.view` are both present. [See here](https://github.com/emberjs/ember.js/blob/dc479cef91c1b9815157612c2a2f9483a2a04edc/packages/ember-htmlbars/lib/hooks/bind-shadow-scope.js#L34)
* But for the components template, we are using our friend `manualElement` and are only passing `view` into the `blockFor` invocation. [See here](https://github.com/emberjs/ember.js/blob/dc479cef91c1b9815157612c2a2f9483a2a04edc/packages/ember-views/lib/system/build-component-template.js#L92-L98)

So it isn't possible that `scope.component` will be set for the components own element (because we were never passing `attrs` in that scenario).

I have updated the code in `bind-shadown-scope` to set `scope.component` if the `view` is stamped with `view.isComponent`.

rwjblue committed:
[BUGFIX release] Allow bindings in `{{link-to}}` invocation.

Any bound properties supplied for things like `current-when` or
`activeClass` would not be properly be used.

This is due to our default of two way bindings. In the case of a bound
CP specified for `activeClass`, `this.attrs.activeClass` would not be a
string, it will be a "Mutable Object".

This commit changes to use `this.getAttr` which automatically unwraps
the mutable object.
